File: apps/web/src/surfaces.ts
import type { ComposeIntent } from './composeIntent'
import type { MailSelection } from './mailState'

export type SurfaceDisposition = 'focused'
export type SettingsSurfaceCategory =
  | 'general'
  | 'appearance'
  | 'accounts'
  | 'mailboxes'

export const SettingsSurfaceTargetKind = {
  Account: 'account',
  NewAccount: 'newAccount',
  SmartMailbox: 'smartMailbox',
  NewSmartMailbox: 'newSmartMailbox',
  SourceMailbox: 'sourceMailbox',
} as const

export type SettingsSurfaceTargetKind =
  (typeof SettingsSurfaceTargetKind)[keyof typeof SettingsSurfaceTargetKind]

export type SettingsSurfaceTarget =
  | { kind: typeof SettingsSurfaceTargetKind.Account; accountId: string }
  | { kind: typeof SettingsSurfaceTargetKind.NewAccount }
  | {
      kind: typeof SettingsSurfaceTargetKind.SmartMailbox
      smartMailboxId: string
    }
  | { kind: typeof SettingsSurfaceTargetKind.NewSmartMailbox }
  | {
      kind: typeof SettingsSurfaceTargetKind.SourceMailbox
      sourceAccountId: string
      sourceMailboxId: string
    }

export interface MessageSurfaceDescriptor {
  kind: 'message'
  disposition: SurfaceDisposition
  params: {
    conversationId: string
    sourceId: string
    messageId: string
  }
}

export interface AttachmentSurfaceDescriptor {
  kind: 'attachment'
  disposition: SurfaceDisposition
  params: {
    sourceId: string
    messageId: string
    attachmentId: string
  }
}

export interface SettingsSurfaceDescriptor {
  kind: 'settings'
  disposition: SurfaceDisposition
  params: {
    category?: SettingsSurfaceCategory
    target?: SettingsSurfaceTarget | null
  }
}

export interface ComposeSurfaceDescriptor {
  kind: 'compose'
  disposition: SurfaceDisposition
  params: ComposeIntent
}

export type SurfaceDescriptor =
  | MessageSurfaceDescriptor
  | AttachmentSurfaceDescriptor
  | SettingsSurfaceDescriptor
  | ComposeSurfaceDescriptor

const SETTINGS_TARGET_PARAM_KEYS = [
  'accountId',
  'smartMailboxId',
  'sourceAccountId',
  'sourceMailboxId',
] as const

type SettingsTargetParamKey = (typeof SETTINGS_TARGET_PARAM_KEYS)[number]

export function messageSurfaceFromSelection(
  selection: MailSelection,
): MessageSurfaceDescriptor {
  return {
    kind: 'message',
    disposition: 'focused',
    params: {
      conversationId: selection.conversationId,
      sourceId: selection.sourceId,
      messageId: selection.messageId,
    },
  }
}

export function settingsSurface(input?: {
  category?: SettingsSurfaceCategory
  target?: SettingsSurfaceTarget | null
}): SettingsSurfaceDescriptor {
  const target = input?.target ?? null
  const category = target ? categoryForSettingsTarget(target) : input?.category
  const params: SettingsSurfaceDescriptor['params'] = {}
  if (category) {
    params.category = category
  }
  if (target) {
    params.target = target
  }
  return {
    kind: 'settings',
    disposition: 'focused',
    params,
  }
}

export function settingsCategorySurface(
  category: SettingsSurfaceCategory,
): SettingsSurfaceDescriptor {
  return settingsSurface({ category })
}

export function accountSettingsSurface(
  accountId: string,
): SettingsSurfaceDescriptor {
  return settingsSurface({
    target: { kind: SettingsSurfaceTargetKind.Account, accountId },
  })
}

export function newAccountSettingsSurface(): SettingsSurfaceDescriptor {
  return settingsSurface({
    target: { kind: SettingsSurfaceTargetKind.NewAccount },
  })
}

export function smartMailboxSettingsSurface(
  smartMailboxId: string,
): SettingsSurfaceDescriptor {
  return settingsSurface({
    target: { kind: SettingsSurfaceTargetKind.SmartMailbox, smartMailboxId },
  })
}

export function newSmartMailboxSettingsSurface(): SettingsSurfaceDescriptor {
  return settingsSurface({
    target: { kind: SettingsSurfaceTargetKind.NewSmartMailbox },
  })
}

export function sourceMailboxSettingsSurface(
  sourceAccountId: string,
  sourceMailboxId: string,
): SettingsSurfaceDescriptor {
  return settingsSurface({
    target: {
      kind: SettingsSurfaceTargetKind.SourceMailbox,
      sourceAccountId,
      sourceMailboxId,
    },
  })
}

export function composeSurface(
  intent: ComposeIntent,
): ComposeSurfaceDescriptor {
  return {
    kind: 'compose',
    disposition: 'focused',
    params: intent,
  }
}

export function attachmentSurface(input: {
  sourceId: string
  messageId: string
  attachmentId: string
}): AttachmentSurfaceDescriptor {
  return {
    kind: 'attachment',
    disposition: 'focused',
    params: input,
  }
}

export function surfaceRoute(surface: SurfaceDescriptor): string {
  switch (surface.kind) {
    case 'settings':
      return settingsSurfaceRoute(surface)
    case 'compose':
      return composeSurfaceRoute(surface)
    case 'message':
    case 'attachment':
      return genericSurfaceRoute(surface)
  }
}

export function parseSurfaceRoute(route: string): SurfaceDescriptor | null {
  const url = new URL(route, 'http://posthaste.local')
  const parts = url.pathname.split('/').filter(Boolean)
  if (parts.length !== 2 || parts[0] !== 'surface') {
    return null
  }

  switch (parts[1]) {
    case 'message': {
      const conversationId = url.searchParams.get('conversationId')
      const sourceId = url.searchParams.get('sourceId')
      const messageId = url.searchParams.get('messageId')
      if (!conversationId || !sourceId || !messageId) {
        return null
      }
      return {
        kind: 'message',
        disposition: 'focused',
        params: { conversationId, sourceId, messageId },
      }
    }
    case 'attachment': {
      const sourceId = url.searchParams.get('sourceId')
      const messageId = url.searchParams.get('messageId')
      const attachmentId = url.searchParams.get('attachmentId')
      if (!sourceId || !messageId || !attachmentId) {
        return null
      }
      return attachmentSurface({ sourceId, messageId, attachmentId })
    }
    case 'compose': {
      const intent = parseComposeIntent(url.searchParams)
      return intent ? composeSurface(intent) : null
    }
    case 'settings': {
      const category = url.searchParams.get('category')
      if (category !== null && !isSettingsSurfaceCategory(category)) {
        return null
      }
      const target = parseSettingsTarget(url.searchParams)
      if (target === undefined) {
        return null
      }
      if (
        target &&
        category &&
        category !== categoryForSettingsTarget(target)
      ) {
        return null
      }
      return settingsSurface({ category: category ?? undefined, target })
    }
    default:
      return null
  }
}

export function surfaceFromLocation(
  location: Location,
): SurfaceDescriptor | null {
  const hashRoute = location.hash.startsWith('#') ? location.hash.slice(1) : ''
  const route =
    hashRoute.length > 0 ? hashRoute : `${location.pathname}${location.search}`
  return parseSurfaceRoute(route)
}

function genericSurfaceRoute(
  surface: MessageSurfaceDescriptor | AttachmentSurfaceDescriptor,
): string {
  const params = new URLSearchParams()
  for (const [key, value] of Object.entries(surface.params)) {
    params.set(key, value)
  }
  const query = params.toString()
  return `/surface/${surface.kind}${query ? `?${query}` : ''}`
}

function composeSurfaceRoute(surface: ComposeSurfaceDescriptor): string {
  const params = new URLSearchParams()
  params.set('composeKind', surface.params.kind)
  params.set('sourceId', surface.params.sourceId)
  if (surface.params.kind !== 'new') {
    params.set('messageId', surface.params.messageId)
  }
  return `/surface/compose?${params.toString()}`
}

function settingsSurfaceRoute(surface: SettingsSurfaceDescriptor): string {
  const params = new URLSearchParams()
  if (surface.params.category) {
    params.set('category', surface.params.category)
  }

  const target = surface.params.target
  if (target) {
    params.set('targetKind', target.kind)
    switch (target.kind) {
      case SettingsSurfaceTargetKind.Account:
        params.set('accountId', target.accountId)
        break
      case SettingsSurfaceTargetKind.NewAccount:
        break
      case SettingsSurfaceTargetKind.SmartMailbox:
        params.set('smartMailboxId', target.smartMailboxId)
        break
      case SettingsSurfaceTargetKind.NewSmartMailbox:
        break
      case SettingsSurfaceTargetKind.SourceMailbox:
        params.set('sourceAccountId', target.sourceAccountId)
        params.set('sourceMailboxId', target.sourceMailboxId)
        break
    }
  }

  const query = params.toString()
  return `/surface/settings${query ? `?${query}` : ''}`
}

function categoryForSettingsTarget(
  target: SettingsSurfaceTarget,
): SettingsSurfaceCategory {
  switch (target.kind) {
    case SettingsSurfaceTargetKind.Account:
    case SettingsSurfaceTargetKind.NewAccount:
      return 'accounts'
    case SettingsSurfaceTargetKind.SmartMailbox:
    case SettingsSurfaceTargetKind.NewSmartMailbox:
    case SettingsSurfaceTargetKind.SourceMailbox:
      return 'mailboxes'
  }
}

function parseComposeIntent(params: URLSearchParams): ComposeIntent | null {
  const composeKind = params.get('composeKind')
  const sourceId = params.get('sourceId')
  if (!composeKind || !sourceId) {
    return null
  }

  switch (composeKind) {
    case 'new':
      return hasOnlyComposeParams(params, []) ? { kind: 'new', sourceId } : null
    case 'reply': {
      const messageId = params.get('messageId')
      return messageId && hasOnlyComposeParams(params, ['messageId'])
        ? { kind: 'reply', sourceId, messageId }
        : null
    }
    case 'forward': {
      const messageId = params.get('messageId')
      return messageId && hasOnlyComposeParams(params, ['messageId'])
        ? { kind: 'forward', sourceId, messageId }
        : null
    }
    default:
      return null
  }
}

function hasOnlyComposeParams(
  params: URLSearchParams,
  expectedKeys: readonly string[],
): boolean {
  return [...params.keys()].every(
    (key) =>
      key === 'composeKind' || key === 'sourceId' || expectedKeys.includes(key),
  )
}

function parseSettingsTarget(
  params: URLSearchParams,
): SettingsSurfaceTarget | null | undefined {
  const targetKind = params.get('targetKind')
  if (targetKind === null) {
    return null
  }

  switch (targetKind) {
    case SettingsSurfaceTargetKind.Account: {
      if (!hasOnlySettingsTargetParams(params, ['accountId'])) {
        return undefined
      }
      const accountId = params.get('accountId')
      return accountId
        ? { kind: SettingsSurfaceTargetKind.Account, accountId }
        : undefined
    }
    case SettingsSurfaceTargetKind.NewAccount:
      if (!hasOnlySettingsTargetParams(params, [])) {
        return undefined
      }
      return { kind: SettingsSurfaceTargetKind.NewAccount }
    case SettingsSurfaceTargetKind.SmartMailbox: {
      if (!hasOnlySettingsTargetParams(params, ['smartMailboxId'])) {
        return undefined
      }
      const smartMailboxId = params.get('smartMailboxId')
      return smartMailboxId
        ? { kind: SettingsSurfaceTargetKind.SmartMailbox, smartMailboxId }
        : undefined
    }
    case SettingsSurfaceTargetKind.NewSmartMailbox:
      if (!hasOnlySettingsTargetParams(params, [])) {
        return undefined
      }
      return { kind: SettingsSurfaceTargetKind.NewSmartMailbox }
    case SettingsSurfaceTargetKind.SourceMailbox: {
      if (
        !hasOnlySettingsTargetParams(params, [
          'sourceAccountId',
          'sourceMailboxId',
        ])
      ) {
        return undefined
      }
      const sourceAccountId = params.get('sourceAccountId')
      const sourceMailboxId = params.get('sourceMailboxId')
      return sourceAccountId && sourceMailboxId
        ? {
            kind: SettingsSurfaceTargetKind.SourceMailbox,
            sourceAccountId,
            sourceMailboxId,
          }
        : undefined
    }
    default:
      return undefined
  }
}

function hasOnlySettingsTargetParams(
  params: URLSearchParams,
  allowed: readonly SettingsTargetParamKey[],
): boolean {
  return SETTINGS_TARGET_PARAM_KEYS.every(
    (key) => allowed.includes(key) || !params.has(key),
  )
}

function isSettingsSurfaceCategory(
  value: string,
): value is SettingsSurfaceCategory {
  return (
    value === 'general' ||
    value === 'appearance' ||
    value === 'accounts' ||
    value === 'mailboxes'
  )
}

File: apps/client/desktop/src/surfaces.rs
//! Surface descriptors and their window plumbing: the wire shapes the
//! frontend posts to `open_surface_window` (mirroring
//! `apps/client/frontend/src/surfaces/types.ts`), the route each surface
//! loads (mirroring `surfaces/serialize.ts`), and the per-surface window
//! label, title, and size.

use serde::Deserialize;

#[derive(Debug, Deserialize)]
#[serde(tag = "kind", rename_all = "camelCase", deny_unknown_fields)]
pub(crate) enum SurfaceDescriptor {
    #[serde(rename = "attachment")]
    Attachment {
        disposition: SurfaceDisposition,
        params: AttachmentSurfaceParams,
    },
    #[serde(rename = "message")]
    Message {
        disposition: SurfaceDisposition,
        params: MessageSurfaceParams,
    },
    #[serde(rename = "settings")]
    Settings {
        disposition: SurfaceDisposition,
        params: SettingsSurfaceParams,
    },
    #[serde(rename = "compose")]
    Compose {
        disposition: SurfaceDisposition,
        params: ComposeSurfaceParams,
    },
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub(crate) enum SurfaceDisposition {
    #[serde(rename = "focused")]
    Focused,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub(crate) struct AttachmentSurfaceParams {
    pub(crate) source_id: String,
    pub(crate) message_id: String,
    pub(crate) attachment_id: String,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub(crate) struct MessageSurfaceParams {
    pub(crate) conversation_id: String,
    pub(crate) source_id: String,
    pub(crate) message_id: String,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub(crate) struct SettingsSurfaceParams {
    pub(crate) category: Option<SettingsSurfaceCategory>,
    pub(crate) target: Option<SettingsSurfaceTarget>,
}

/// Mirrors `SETTINGS_SURFACE_CATEGORIES` in the frontend's surface types.
#[derive(Debug, Clone, Copy, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub(crate) enum SettingsSurfaceCategory {
    General,
    Appearance,
    Accounts,
    Outbox,
    Mailboxes,
    Automations,
    Tags,
    Storage,
    Notifications,
    Troubleshooting,
}

impl SettingsSurfaceCategory {
    pub(crate) fn as_str(&self) -> &'static str {
        match self {
            Self::General => "general",
            Self::Appearance => "appearance",
            Self::Accounts => "accounts",
            Self::Outbox => "outbox",
            Self::Mailboxes => "mailboxes",
            Self::Automations => "automations",
            Self::Tags => "tags",
            Self::Storage => "storage",
            Self::Notifications => "notifications",
            Self::Troubleshooting => "troubleshooting",
        }
    }
}

/// Mirrors `ComposeIntent` in `apps/client/frontend/src/composeIntent.ts`.
#[derive(Debug, Deserialize)]
#[serde(tag = "kind", rename_all = "camelCase", deny_unknown_fields)]
pub(crate) enum ComposeSurfaceParams {
    #[serde(rename = "new")]
    New {
        #[serde(rename = "sourceId")]
        source_id: String,
    },
    #[serde(rename = "reply")]
    Reply {
        #[serde(rename = "sourceId")]
        source_id: String,
        #[serde(rename = "messageId")]
        message_id: String,
    },
    #[serde(rename = "replyAll")]
    ReplyAll {
        #[serde(rename = "sourceId")]
        source_id: String,
        #[serde(rename = "messageId")]
        message_id: String,
    },
    #[serde(rename = "forward")]
    Forward {
        #[serde(rename = "sourceId")]
        source_id: String,
        #[serde(rename = "messageId")]
        message_id: String,
    },
    #[serde(rename = "draft")]
    Draft {
        #[serde(rename = "sourceId")]
        source_id: String,
        #[serde(rename = "messageId")]
        message_id: String,
    },
    #[serde(rename = "mailto")]
    Mailto {
        #[serde(rename = "sourceId")]
        source_id: String,
        #[serde(rename = "mailtoUri")]
        mailto_uri: String,
    },
}

impl ComposeSurfaceParams {
    fn kind_str(&self) -> &'static str {
        match self {
            Self::New { .. } => "new",
            Self::Reply { .. } => "reply",
            Self::ReplyAll { .. } => "replyAll",
            Self::Forward { .. } => "forward",
            Self::Draft { .. } => "draft",
            Self::Mailto { .. } => "mailto",
        }
    }

    fn source_id(&self) -> &str {
        match self {
            Self::New { source_id }
            | Self::Reply { source_id, .. }
            | Self::ReplyAll { source_id, .. }
            | Self::Forward { source_id, .. }
            | Self::Draft { source_id, .. }
            | Self::Mailto { source_id, .. } => source_id,
        }
    }
}

#[derive(Debug, Deserialize)]
#[serde(tag = "kind", rename_all = "camelCase", deny_unknown_fields)]
pub(crate) enum SettingsSurfaceTarget {
    #[serde(rename = "account")]
    Account {
        #[serde(rename = "accountId")]
        account_id: String,
    },
    #[serde(rename = "newAccount")]
    NewAccount,
    #[serde(rename = "smartMailbox")]
    SmartMailbox {
        #[serde(rename = "smartMailboxId")]
        smart_mailbox_id: String,
    },
    #[serde(rename = "newSmartMailbox")]
    NewSmartMailbox,
    #[serde(rename = "sourceMailbox")]
    SourceMailbox {
        #[serde(rename = "sourceAccountId")]
        source_account_id: String,
        #[serde(rename = "sourceMailboxId")]
        source_mailbox_id: String,
    },
}

/// The hash route a surface loads, mirroring `surfaceRoute` in the
/// frontend's `surfaces/serialize.ts` (same paths, same query keys, same
/// pair order) so a Rust-built route parses identically on the other side.
pub(crate) fn surface_route(surface: &SurfaceDescriptor) -> String {
    match surface {
        SurfaceDescriptor::Attachment {
            disposition,
            params,
        } => {
            let _ = disposition;
            format!(
                "/surface/attachment?sourceId={}&messageId={}&attachmentId={}",
                encode_component(&params.source_id),
                encode_component(&params.message_id),
                encode_component(&params.attachment_id)
            )
        }
        SurfaceDescriptor::Message {
            disposition,
            params,
        } => {
            let _ = disposition;
            format!(
                "/surface/message?conversationId={}&sourceId={}&messageId={}",
                encode_component(&params.conversation_id),
                encode_component(&params.source_id),
                encode_component(&params.message_id)
            )
        }
        SurfaceDescriptor::Settings {
            disposition,
            params,
        } => {
            let _ = disposition;
            let mut pairs = Vec::new();
            push_query_pair(
                &mut pairs,
                "category",
                params
                    .category
                    .as_ref()
                    .map(SettingsSurfaceCategory::as_str),
            );
            push_settings_target_query_pairs(&mut pairs, params.target.as_ref());
            if pairs.is_empty() {
                "/surface/settings".to_string()
            } else {
                format!("/surface/settings?{}", pairs.join("&"))
            }
        }
        SurfaceDescriptor::Compose {
            disposition,
            params,
        } => {
            let _ = disposition;
            let mut pairs = Vec::new();
            push_compose_query_pairs(&mut pairs, params);
            format!("/surface/compose?{}", pairs.join("&"))
        }
    }
}

/// The URL a standalone surface window loads. Always the one real bundled
/// document (`index.html`) with the surface route — query params included —
/// carried in the URL FRAGMENT, so every window on every platform loads the
/// same file and never depends on SPA path-fallback behavior in the asset
/// protocol. The client reads the route (and its query) from `location.hash`.
pub(crate) fn surface_window_url(surface: &SurfaceDescriptor) -> String {
    format!("index.html#{}", surface_route(surface))
}

pub(crate) fn surface_window_navigation_script(route: &str) -> String {
    let route_json = serde_json::to_string(route).expect("surface route should serialize to JSON");
    format!(
        "(() => {{ const route = {route_json}; window.history.replaceState(window.history.state, '', '#' + route); window.dispatchEvent(new HashChangeEvent('hashchange')); }})();"
    )
}

pub(crate) fn validate_surface_descriptor(surface: &SurfaceDescriptor) -> Result<(), String> {
    if let SurfaceDescriptor::Settings { params, .. } = surface {
        if let (Some(category), Some(target)) = (&params.category, &params.target) {
            let target_category = settings_target_category(target);
            if *category != target_category {
                return Err("settings surface category does not match target kind".to_string());
            }
        }
    }
    Ok(())
}

pub(crate) fn settings_target_category(target: &SettingsSurfaceTarget) -> SettingsSurfaceCategory {
    match target {
        SettingsSurfaceTarget::Account { .. } | SettingsSurfaceTarget::NewAccount => {
            SettingsSurfaceCategory::Accounts
        }
        SettingsSurfaceTarget::SmartMailbox { .. }
        | SettingsSurfaceTarget::NewSmartMailbox
        | SettingsSurfaceTarget::SourceMailbox { .. } => SettingsSurfaceCategory::Mailboxes,
    }
}

/// A stable per-surface window label: reopening the same message, attachment,
/// or compose intent focuses the existing window instead of stacking a new
/// one; settings is a singleton.
pub(crate) fn surface_window_label(surface: &SurfaceDescriptor) -> String {
    match surface {
        SurfaceDescriptor::Attachment { params, .. } => {
            let key = format!(
                "{}:{}:{}",
                params.source_id, params.message_id, params.attachment_id
            );
            format!("attachment-{:016x}", stable_hash(key.as_bytes()))
        }
        SurfaceDescriptor::Settings { .. } => "settings".to_string(),
        SurfaceDescriptor::Compose { .. } => {
            format!(
                "compose-{:016x}",
                stable_hash(surface_route(surface).as_bytes())
            )
        }
        SurfaceDescriptor::Message { params, .. } => {
            let key = format!("{}:{}", params.source_id, params.message_id);
            format!("message-{:016x}", stable_hash(key.as_bytes()))
        }
    }
}

pub(crate) fn surface_title(surface: &SurfaceDescriptor) -> &'static str {
    match surface {
        SurfaceDescriptor::Attachment { .. } => "Posthaste Attachment",
        SurfaceDescriptor::Settings { .. } => "Posthaste Settings",
        SurfaceDescriptor::Message { .. } => "Posthaste Message",
        SurfaceDescriptor::Compose { .. } => "Posthaste Compose",
    }
}

/// Window sizes, kept in sync with `surfaceWindowPolicy.ts` popup sizes so
/// the desktop and browser variants of a surface open at the same geometry.
pub(crate) fn surface_window_size(surface: &SurfaceDescriptor) -> (f64, f64) {
    match surface {
        SurfaceDescriptor::Attachment { .. } => (1100.0, 820.0),
        SurfaceDescriptor::Settings { .. } => (980.0, 720.0),
        SurfaceDescriptor::Message { .. } => (900.0, 760.0),
        SurfaceDescriptor::Compose { .. } => (780.0, 640.0),
    }
}

fn push_query_pair(pairs: &mut Vec<String>, key: &str, value: Option<&str>) {
    if let Some(value) = value {
        pairs.push(format!("{key}={}", encode_component(value)));
    }
}

fn push_compose_query_pairs(pairs: &mut Vec<String>, params: &ComposeSurfaceParams) {
    push_query_pair(pairs, "composeKind", Some(params.kind_str()));
    push_query_pair(pairs, "sourceId", Some(params.source_id()));
    match params {
        ComposeSurfaceParams::New { .. } => {}
        ComposeSurfaceParams::Reply { message_id, .. }
        | ComposeSurfaceParams::ReplyAll { message_id, .. }
        | ComposeSurfaceParams::Forward { message_id, .. }
        | ComposeSurfaceParams::Draft { message_id, .. } => {
            push_query_pair(pairs, "messageId", Some(message_id));
        }
        ComposeSurfaceParams::Mailto { mailto_uri, .. } => {
            push_query_pair(pairs, "mailtoUri", Some(mailto_uri));
        }
    }
}

fn push_settings_target_query_pairs(
    pairs: &mut Vec<String>,
    target: Option<&SettingsSurfaceTarget>,
) {
    let Some(target) = target else {
        return;
    };

    match target {
        SettingsSurfaceTarget::Account { account_id } => {
            push_query_pair(pairs, "targetKind", Some("account"));
            push_query_pair(pairs, "accountId", Some(account_id));
        }
        SettingsSurfaceTarget::NewAccount => {
            push_query_pair(pairs, "targetKind", Some("newAccount"));
        }
        SettingsSurfaceTarget::SmartMailbox { smart_mailbox_id } => {
            push_query_pair(pairs, "targetKind", Some("smartMailbox"));
            push_query_pair(pairs, "smartMailboxId", Some(smart_mailbox_id));
        }
        SettingsSurfaceTarget::NewSmartMailbox => {
            push_query_pair(pairs, "targetKind", Some("newSmartMailbox"));
        }
        SettingsSurfaceTarget::SourceMailbox {
            source_account_id,
            source_mailbox_id,
        } => {
            push_query_pair(pairs, "targetKind", Some("sourceMailbox"));
            push_query_pair(pairs, "sourceAccountId", Some(source_account_id));
            push_query_pair(pairs, "sourceMailboxId", Some(source_mailbox_id));
        }
    }
}

fn encode_component(value: &str) -> String {
    let mut encoded = String::new();
    for byte in value.bytes() {
        match byte {
            b'A'..=b'Z' | b'a'..=b'z' | b'0'..=b'9' | b'-' | b'_' | b'.' | b'~' => {
                encoded.push(byte as char);
            }
            _ => encoded.push_str(&format!("%{byte:02X}")),
        }
    }
    encoded
}

fn stable_hash(bytes: &[u8]) -> u64 {
    let mut hash = 0xcbf29ce484222325_u64;
    for byte in bytes {
        hash ^= u64::from(*byte);
        hash = hash.wrapping_mul(0x100000001b3);
    }
    hash
}
